hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/hudi/common/TestTableColumnTypeMismatch.scala

```diff
@@ -0,0 +1,919 @@
+      }
+    })
+  }
+
+  test("Test All Valid Type Casting For Merge Into and Insert") {
+    // For all valid type casting pairs, test merge into and insert operations.
+    // Define the column types for testing, based on successful casting cases
+    case class ColumnTypePair(
+                               sourceType: String,
+                               targetType: String,
+                               testValue: String,
+                               expectedValue: Any,
+                               columnName: String
+                             )
+
+    // Define valid type casting pairs based on the previous test cases
+    val validTypePairs = Seq(
+      // Numeric widening pairs
+      ColumnTypePair("tinyint", "smallint", "127", 127, "tiny_to_small"),
+      ColumnTypePair("tinyint", "int", "127", 127, "tiny_to_int"),
+      ColumnTypePair("tinyint", "bigint", "127", 127L, "tiny_to_big"),
+      ColumnTypePair("tinyint", "float", "127", 127.0f, "tiny_to_float"),
+      ColumnTypePair("tinyint", "double", "127", 127.0d, "tiny_to_double"),
+      ColumnTypePair("tinyint", "decimal(10,1)", "127", java.math.BigDecimal.valueOf(127.0), "tiny_to_decimal"),
+
+      ColumnTypePair("smallint", "int", "32767", 32767, "small_to_int"),
+      ColumnTypePair("smallint", "bigint", "32767", 32767L, "small_to_big"),
+      ColumnTypePair("smallint", "float", "32767", 32767.0f, "small_to_float"),
+      ColumnTypePair("smallint", "double", "32767", 32767.0d, "small_to_double"),
+      ColumnTypePair("smallint", "decimal(10,1)", "32767", java.math.BigDecimal.valueOf(32767.0), "small_to_decimal"),
+
+      ColumnTypePair("int", "bigint", "2147483647", 2147483647L, "int_to_big"),
+      ColumnTypePair("int", "float", "2147483647", 2147483647.0f, "int_to_float"),
+      ColumnTypePair("int", "double", "2147483647", 2147483647.0d, "int_to_double"),
+      ColumnTypePair("int", "decimal(10,1)", "22", java.math.BigDecimal.valueOf(22.0), "int_to_decimal"),
+
+      ColumnTypePair("float", "double", "3.14", 3.140000104904175d, "float_to_double"),
+      ColumnTypePair("float", "decimal(10,2)", "3.14", java.math.BigDecimal.valueOf(3.14).setScale(2, java.math.RoundingMode.HALF_UP), "float_to_decimal"),
+
+      // Timestamp/Date conversions
+      ColumnTypePair("timestamp", "string", "timestamp'2023-01-01 12:00:00'", "2023-01-01 12:00:00", "ts_to_string"),
+      ColumnTypePair("timestamp", "date", "timestamp'2023-01-01 12:00:00'", java.sql.Date.valueOf("2023-01-01"), "ts_to_date"),
+      ColumnTypePair("date", "string", "date'2023-01-01'", "2023-01-01", "date_to_string"),
+      ColumnTypePair("date", "timestamp", "date'2023-01-01'", java.sql.Timestamp.valueOf("2023-01-01 00:00:00"), "date_to_ts"),
+
+      // Boolean conversions
+      ColumnTypePair("boolean", "string", "true", "true", "bool_to_string")
+    )
+
+    Seq("cow", "mor").foreach { tableType =>
+      withRecordType()(withTempDir { tmp =>
```

Comment:
I removed the permutation on the record types as they are irrelevant to the tests here.